File: binary_search/binary_search.ex
defmodule Algorithms do
  def binary_search(tup, item) do
    low = 0
    high = tuple_size(tup) - 1
    _mid = do_search(tup, item, low, high)
  end

  defp do_search(_tup, _item, :found, mid), do: mid
  defp do_search(_tup, _item, low, high) when low > high, do: -1

  defp do_search(tup, item, low, high) when low <= high do
    mid = div(low + high, 2)
    guess = elem(tup, mid)

    {low, mid} =
      cond do
        guess == item -> {:found, mid}
        item < guess -> {low, mid - 1}
        item > guess -> {mid + 1, high}
      end

    do_search(tup, item, low, mid)
  end
end

File: rust/src/binary_search.rs
pub fn binary_search<T: Ord>(arr: &[T], item: &T) -> Option<usize> {
    let mut low = 0;
    let mut high = arr.len() - 1;

    while low <= high {
        let mid = low + (high - low) / 2;
        let guess = &arr[mid];
        if *guess == *item {
            return Some(mid);
        }
        if *item < *guess {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    None
}
